controllers/api_v1/login_test.go

```diff
@@ -0,0 +1,172 @@
+		ID:     "client-id",
+		Secret: "client-secret",
+		Domain: "http://localhost:8080",
+	}
+	data, _ := json.Marshal(s.oauthClient)
+	s.oauthClient.Data = data
+
+	db.GetDB().Exec("INSERT INTO oauth2_clients VALUES(?, ?, ?, ?)",
+		s.oauthClient.ID,
+		s.oauthClient.Secret,
+		s.oauthClient.Domain,
+		s.oauthClient.Data,
+	)
+}
+
+func (s *LoginAPIControllerDbTestSuite) TearDownTest() {
+	db.GetDB().Exec("DELETE FROM users")
+	db.GetDB().Exec("DELETE FROM oauth2_clients")
+	db.GetDB().Exec("DELETE FROM oauth2_tokens")
+}
+
+func TestLoginAPIControllerDbTestSuite(t *testing.T) {
+	suite.Run(t, new(LoginAPIControllerDbTestSuite))
+}
+
+func (s *LoginAPIControllerDbTestSuite) TestGenerateTokenWithValidParams() {
+	formData := url.Values{}
+	formData.Set("username", s.user.Email)
+	formData.Set("password", "password")
+	formData.Set("grant_type", "password")
+	formData.Set("client_id", s.oauthClient.ID)
+	formData.Set("client_secret", s.oauthClient.Secret)
+
+	resp := testHttp.PerformRequest(s.engine, "POST", "/api/v1/login", s.headers, formData)
+	respBodyData, _ := ioutil.ReadAll(resp.Body)
+	var parsedRespBody map[string]string
+	_ = json.Unmarshal(respBodyData, &parsedRespBody)
+
+	var data []byte
+	row := db.GetDB().Table("oauth2_tokens").Select("data").Row()
+	_ = row.Scan(&data)
+
+	var dataVal map[string]interface{}
+	_ = json.Unmarshal(data, &dataVal)
+
+	assert.Equal(s.T(), http.StatusOK, resp.Code)
+	assert.Equal(s.T(), parsedRespBody["access_token"], dataVal["Access"])
+	assert.Equal(s.T(), parsedRespBody["refresh_token"], dataVal["Refresh"])
+	assert.Equal(s.T(), fmt.Sprint(s.user.ID), dataVal["UserID"])
+}
+
+func (s *LoginAPIControllerDbTestSuite) TestGenerateTokenWithInvalidGrantType() {
+	formData := url.Values{}
+	formData.Set("username", s.user.Email)
+	formData.Set("password", "password")
+	formData.Set("grant_type", "invalidGrant")
+	formData.Set("client_id", s.oauthClient.ID)
+	formData.Set("client_secret", s.oauthClient.Secret)
+
+	resp := testHttp.PerformRequest(s.engine, "POST", "/api/v1/login", s.headers, formData)
+	respBodyData, _ := ioutil.ReadAll(resp.Body)
+	var parsedRespBody map[string]string
+	_ = json.Unmarshal(respBodyData, &parsedRespBody)
+
+	assert.Equal(s.T(), http.StatusUnauthorized, resp.Code)
+	assert.Equal(s.T(), errors.ErrUnsupportedGrantType.Error(), parsedRespBody["error"])
+}
+
+func (s *LoginAPIControllerDbTestSuite) TestGenerateTokenWithInvalidClientID() {
+	formData := url.Values{}
+	formData.Set("username", s.user.Email)
+	formData.Set("password", "password")
+	formData.Set("grant_type", "password")
+	formData.Set("client_id", "invalid")
+	formData.Set("client_secret", s.oauthClient.Secret)
+
+	resp := testHttp.PerformRequest(s.engine, "POST", "/api/v1/login", s.headers, formData)
+	respBodyData, _ := ioutil.ReadAll(resp.Body)
+	var parsedRespBody map[string]string
+	_ = json.Unmarshal(respBodyData, &parsedRespBody)
+
+	assert.Equal(s.T(), http.StatusInternalServerError, resp.Code)
```

Comment:
`StatusInternalServerError ` seems weird for this case, should it be `StatusUnauthorized` ?

Reply:
It is the error return from `go-oauth2`. Same as above if we need to do it ourselves and it cause a lot of extra work.

Reply:
It would definitely need to be implemented as `StatusUnauthorized` so please add a feature task for this :-)

Reply:
Added TODO in 89619b88287370528ea12fc6c21e9349c18fd59a and created new issue in https://github.com/gutakk/go-google-scraper/issues/72